pool_impl.go

```diff
@@ -0,0 +1,113 @@
+package goconnpool
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+)
+
+var (
+	ErrNoServersRegistered = fmt.Errorf("no registered servers found")
+)
+
+type connPool struct {
+	cfg Config
+
+	mu sync.Mutex
+
+	servers             roundRobin
+	connProviderFactory func(addr string, cfg Config) connectionProvider
+}
+
+func newConnPool(cfg Config) *connPool {
+	cfg = cfg.withDefaults()
+
+	return &connPool{
+		cfg:                 cfg,
+		connProviderFactory: newServerWrapper, // required for tests
+	}
+}
+
+func (p *connPool) OpenConn(ctx context.Context) (Conn, error) {
+	for {
```

Comment:
Forever till !ctx.Done()? Don't you want to use some timeout, when everything really drops or some 'redialForever' option?

Reply:
For what?
Timeout could be made with context.WithTimeout() wrapper.

Reply:
hm, okay
mb its better to create helpers like 
```OpenConnWithTimeout(ctx context.Context, tm time.Duration) (Conn, error) { 
ctx1, cancel := context.WithTimeout(ctx, tm)
defer cancel()
return OpenConn(ctx1)
} 
```

Reply:
hmmm
makes sense

Reply:
Fixed